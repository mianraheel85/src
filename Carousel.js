import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";

/** Carousel: displays images and arrows to navigate through them
 *
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 *
 * State:
 * - currCardIdx: integer for current card index
 *
 * App --> Carousel --> Card
 */
function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);

  const currCard = photos[currCardIdx];
  const total = photos.length;

  //Increments currCardIdx state by 1
  function goForward() {
    setCurrCardIdx(currCardIdx + 1);
  }

  //Decrement currCardIdx state by 1
  function goBackward() {
    setCurrCardIdx(currCardIdx - 1);
  }

  return (
    <div className="Carousel">
      <h1>{title}</h1>
      {/* only render photos if there are photos */}
      {total > 0 ? (
        <div className="Carousel-main">
          {/* Hide left arrow on the first image */}
          {currCardIdx > 0 && (
            <i className="bi bi-arrow-left-circle" onClick={goBackward} />
          )}
          <Card
            caption={currCard.caption}
            src={currCard.src}
            currNum={currCardIdx + 1}
            totalNum={total}
          />
          {/* Hide right arrow on the last image */}
          {currCardIdx < total - 1 && (
            <i className="bi bi-arrow-right-circle" onClick={goForward} />
          )}
        </div>
      ) : (
        <p>No photos</p>
      )}
    </div>
  );
}

export default Carousel;
