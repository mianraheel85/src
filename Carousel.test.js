import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";

//  smoke test
test("Carousel renders without crashing", () => {
  render(<Carousel photos={[]} title="Test Carousel" />);
  // If the component renders without throwing an error, the test passes
});

// snapshot test
test("Carousel snapshot test", () => {
  const { asFragment } = render(<Carousel photos={[]} title="Test Carousel" />);
  // Capture the rendered output
  const component = asFragment();
  // Compare with the saved snapshot
  expect(component).toMatchSnapshot();
});

// clicking on left arrow
it("works when you click on the left arrow", function () {
  const { container } = render(
    <Carousel photos={TEST_IMAGES} title="images for testing" />
  );
  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();

  // move to image 2
  fireEvent.click(container.querySelector(".bi-arrow-right-circle"));

  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).toBeInTheDocument();

  // move backward in the carousel
  const leftArrow = container.querySelector(".bi-arrow-left-circle");
  fireEvent.click(leftArrow);

  // expect the first image to show
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
});

// clicking on right arrow
it("works when you click on the right arrow", function () {
  const { container } = render(
    <Carousel photos={TEST_IMAGES} title="images for testing" />
  );
  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).not.toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).toBeInTheDocument();
});

// missing arrows
test("left arrow is missing on the first image, and right arrow is missing on the last image", () => {
  const { container } = render(
    <Carousel photos={TEST_IMAGES} title="Test Carousel" />
  );

  // Check that the left arrow is missing on the first image
  const leftArrow = container.querySelector(".bi-arrow-left-circle");
  expect(leftArrow).not.toBeInTheDocument(); // This will fail

  const rightArrow = container.querySelector(".bi-arrow-right-circle");

  // Move to the last image
  TEST_IMAGES.forEach(() => {
    //expect(rightArrow).toBeInTheDocument();
    fireEvent.click(rightArrow);
  });

  // Check that the right arrow is missing on the last image
  // const rightArrow = container.querySelector(".bi-arrow-right-circle");
  expect(rightArrow).not.toBeInTheDocument(); // This will fail
});
