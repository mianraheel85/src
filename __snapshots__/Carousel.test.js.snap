// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel snapshot test 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Test Carousel
    </h1>
    <p>
      No photos
    </p>
  </div>
</DocumentFragment>
`;
